src/stdlib_sparse_kinds.fypp

```diff
@@ -0,0 +1,420 @@
+            allocate(temp_idx(num_cols+1) , source = 0 )
+        else
+            allocate(temp_idx(num_cols+1) , source = self%colptr )
+        end if
+        call move_alloc(from=temp_idx,to=self%colptr)
+
+        select type(self)
+            #:for k1, t1, s1 in (KINDS_TYPES)
+            type is(CSC_${s1}$)
+                block
+                ${t1}$, allocatable :: temp_data_${s1}$(:)
+                if(.not.allocated(self%data)) then
+                    allocate(temp_data_${s1}$(nnz) , source = zero_${s1}$ )
+                else
+                    allocate(temp_data_${s1}$(nnz) , source = self%data )
+                end if
+                call move_alloc(from=temp_data_${s1}$,to=self%data)
+                end block
+            #:endfor
+        end select
+    end subroutine
+
+    !> (re)Allocate matrix memory for the ELLPACK type
+    subroutine malloc_ell(self,num_rows,num_cols,num_nz_rows)
+        class(ELL_type) :: self
+        integer, intent(in) :: num_rows    !> number of rows
+        integer, intent(in) :: num_cols    !> number of columns
+        integer, intent(in) :: num_nz_rows !> number of non zeros per row
+
+        integer,  allocatable :: temp_idx(:,:)
+        !-----------------------------------------------------
+
+        self%nrows = num_rows
+        self%ncols = num_cols
+        self%K     = num_nz_rows
+
+        if(.not.allocated(self%index)) then
+            allocate(temp_idx(num_rows,num_nz_rows) , source = 0 )
+        else
+            allocate(temp_idx(num_rows,num_nz_rows) , source = self%index )
+        end if
+        call move_alloc(from=temp_idx,to=self%index)
+
+        select type(self)
+            #:for k1, t1, s1 in (KINDS_TYPES)
+            type is(ELL_${s1}$)
+                block
+                ${t1}$, allocatable :: temp_data_${s1}$(:,:)
+                if(.not.allocated(self%data)) then
+                    allocate(temp_data_${s1}$(num_rows,num_nz_rows) , source = zero_${s1}$ )
+                else
+                    allocate(temp_data_${s1}$(num_rows,num_nz_rows) , source = self%data )
+                end if
+                call move_alloc(from=temp_data_${s1}$,to=self%data)
+                end block
+            #:endfor
+        end select
+    end subroutine
+
+    !==================================================================
+    ! data accessors
+    !==================================================================
+
+    #:for k1, t1, s1 in (KINDS_TYPES)
+    pure subroutine get_value_coo_${s1}$(self,val,ik,jk)
+        class(COO_${s1}$), intent(in) :: self
+        ${t1}$, intent(out) :: val
+        integer, intent(in) :: ik, jk
+        integer :: k
+        ! naive implementation
+        do k = 1,self%nnz
+            if( ik == self%index(1,k) .and. jk == self%index(2,k) ) then
+                val = self%data(k)
+                return
+            end if
+        end do
+        val = zero_${s1}$
```

Comment:
I like the fact that get/set should all be `pure` subroutines. I suggest we add one more option, that is when the required indices are out of bounds. I think the best case scenario would be: 

- `(i,j)` in matrix bounds -> return value, or 0.0 if not part of the sparse pattern
- `(i,j)` out of bounds -> return `ieee_quiet_nan` 

I think this would give more safety to the users, as it is easy to spot whether wrong inputs were provided. Sparse matrices are fast but not very flexible, so I think the option that matrix size (nrows, ncols) grows dynamically with user-data setting should not be considered.

Reply:
Resolved in https://github.com/fortran-lang/stdlib/pull/760/commits/21a8547d2eec103fe04a6a825b8994a4cac71c5d and https://github.com/fortran-lang/stdlib/pull/760/commits/2fb4e83c3835c4a2ef67c15dc0a6a7112c132c59